flips/20211208-cadence-storage-api-changes.md

```diff
@@ -0,0 +1,86 @@
+# Title of FLIP
+
+| Status        | (Proposed)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [718](https://github.com/onflow/flow/pull/718) (update when you have PR #)|
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2021-12-08                                           |
+
+## Objective
+
+This change would cause the `load`, `borrow`, `copy` and capability borrow 
+methods of accounts to cause a force type-cast between the type of the
+value in storage and the supplied type argument. In effect, this would
+abort execution in cases where the value the user is trying to load or
+borrow does not match the type supplied. 
+
+## Motivation
+
+The load, copy, and borrow functions, as well as the capability borrow function, 
+currently return `nil` both when the storage location is empty, and when the storage 
+location contains a value, but does not have the requested type. The inability 
+to differentiate the two cases makes debugging hard.
+
+## User Benefit
+
+This will result in less ambiguity between the two cases in which a load can fail,
+meaning users will not need to guess whether or not the value did not exist or 
+whether the type argument was wrong. 
+
+## Design Proposal
+
+This changes the interpreter's handling of storage accesses. Previously, if
+the supplied type argument did not match the value present in storage, the 
+method would return `nil`. Now, instead, it will raise a `ForceCastTypeMismatchError`
+between the expected type (the supplied type argument) and the dynamic type
+of the value. 
+
+Concretely, where this code would previously have produced `nil` for the value of `v`:
+```
+account.save(3, to: /storage/path)
+let v = account.load<String>(from: /storage/path/)
+```
+after this change, it will now abort execution.
+
+### Drawbacks
+
+This is a breaking change, and will require existing contracts to be updated to 
```

Comment:
If I understand the change correctly, the code would still work, but if an incorrect type was at the target location the script would panic instead of returning nil.

This to me does not seem to be such a big change. I would say that in general (in my un-researched opinion) developer do not rely on these methods returning nil if a different type is stored at the target location. I imagine that in general this will have no effect, but in some cases it will have the beneficial effect of detecting undesired behaviour.

Reply:
This will make it better then what it is today. The most frequent way this occurs IMHO is when people use the same contract under different addresses. 

The error message will tell what contract the type is from if it is from userland?

Reply:
I agree that it is unlikely to significantly affect any contracts, but developers will have to audit their contracts to make sure they don't crash as a result of the change.

Reply:
>  but developers will have to audit their contracts to make sure they don't crash as a result of the change

Agreed! I was jus wondering if it is worth reiterating in this chapter, that while it is a breaking change, and you should re-check your contract, it is likely that no action (other than re-checking) will be needed.